rskj-core/src/main/java/org/ethereum/crypto/signature/Secp256k1ServiceNative.java

```diff
@@ -19,10 +19,88 @@
 
 package org.ethereum.crypto.signature;
 
+import org.bitcoin.NativeSecp256k1;
+import org.bitcoin.NativeSecp256k1Exception;
+import org.ethereum.crypto.ECKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import static java.lang.System.arraycopy;
+
 /**
  * Implementation of SignatureService with Native library.
- * TODO: once integrated native implementation, should implement all methods.
  */
 public class Secp256k1ServiceNative extends Secp256k1ServiceBC {
 
+    private static final Logger logger = LoggerFactory.getLogger(Secp256k1ServiceNative.class);
+    private static final byte[] ZERO_PUB = {0};
+
+    @Nullable
+    @Override
+    public ECKey recoverFromSignature(int recId, ECDSASignature sig, byte[] messageHash, boolean compressed) {
+        check(recId >= 0, "recId must be positive");
+        check(sig.getR().signum() >= 0, "r must be positive");
+        check(sig.getS().signum() >= 0, "s must be positive");
+        check(messageHash != null, "messageHash must not be null");
+        byte[] pbKey;
+        try {
+            byte[] sigBytes = concatenate(sig);
+            logger.trace("Recovering key from signature: comporessed[{}] - recId[{}] - sig[{}] - msgHash[{}].", compressed, recId, sigBytes, messageHash);
+            try {
+                pbKey = NativeSecp256k1.ecdsaRecover(sigBytes, messageHash, recId, compressed);
+            } catch (NativeSecp256k1Exception e) {
+                if (NativeSecp256k1.isInfinity(sigBytes, messageHash, recId)) {
+                    return ECKey.fromPublicOnly(ZERO_PUB);
+                }
+                throw e;
+            }
+        } catch (Exception e) {
+            logger.error("Couldnt recover key from signature.", e);
+            return null;
+        }
+        return ECKey.fromPublicOnly(pbKey);
+    }
+
+    /**
+     * Returns a (r.length + s.length) bytes array long
+     *
+     * Note: we take 32 bytes from "r" and 32 bytes from "s".
+     *
+     * @param sig {r,s}
+     * @return r + s (64 length byte array)
+     */
+    byte[] concatenate(ECDSASignature sig) {
```

Comment:
This method is ugly when r and s are longer than 32 bytes. If we prepend garbage before the real r and s values, those bytes will be ignored. This does not happen with the BC implementation.

There's no immediate issue with that because the Transaction class does an extra validation on the r and s components, but those are not the only places where the recover from signature is used, causing some ugly cases there.

I think that we should fail quickly here. If r or s are too big.

Reply:
fixed.